src/main/java/com/google/sps/Scheduler.java

```diff
@@ -1,15 +1,113 @@
 
 /**
- * Static class for permuting schedules given a list of courses.
+ * Static class for permuting schedules given a list of courses and a set of Invariants.
  */
 public final class Scheduler {
-    public Collection<Schedule> generateSchedules(Collection<Course> courses, Invariants invariants){
-        //TODO: This
-        return null;
+  /**
+   * Generates all valid schedules given a list of courses and invariant
+   * constraints.
+   * 
+   * @param courses    The list of courses to generate the schedules from.
+   * @param invariants The invariants that constrain the valid schedules
+   * @return A list of valid schedules.
+   */
+  public List<Schedule> generateSchedules(List<Course> courses, Invariants invariants) {
+    ArrayList<Course> betterCourses = new ArrayList<>(courses);
+    Collections.sort(betterCourses, Course.REQUIRED_TO_FRONT);
+    Set<List<Course>> schedulesSet = new HashSet<List<Course>>();
+
+    generateSchedulesHelper(betterCourses, new ArrayList<>(), invariants, schedulesSet);
+
+    System.out.println(courses);
+    List<Schedule> schedules = new ArrayList<>();
+    for (List<Course> schedule : schedulesSet) {
+      schedules.add(new Schedule(schedule));
+    }
+
+    return schedules;
+  }
+
+  /**
+   * Recursively iterates through a list of courses and adds schedules to the
+   * given set.
+   * 
+   * @param courses    The list of courses to iterate through
+   * @param schedule   The current schedule to be added/operated on
+   * @param invariants The invariants that constrain the valid schedules
+   * @param schedules  The set of schedules that is being generated/added to.
+   */
+  private void generateSchedulesHelper(List<Course> courses, List<Course> schedule, 
+      Invariants invariants, Set<List<Course>> schedules) {
+
+    // This means that we have a valid schedule, and should add it to the list.
+    float currCredits = totalCredits(schedule);
+    if (currCredits >= invariants.getMinCredits() && currCredits <= invariants.getMaxCredits()) {
+      schedules.add(schedule);
+    }
+
+    // We've reached the end of the list or have a full schedule, we are finished.
+    if (courses.isEmpty() || currCredits >= invariants.getMaxCredits()) {
+      return;
+    }
+
+    Course course = courses.remove(0);
+    List<Section> courseSections = course.getSections();
+    List<Section> scheduleSections = schedule.stream()
+        .map(c -> c.getSections().get(0))
+        .collect(Collectors.toList());
+
+    /*
+     * Iterate over each section in the current course and check for overlap. If the
+     * section doesn't overlap and adding it would result in a valid schedule, add
+     * it and make a new recursive call.
+     */
+    // TODO: Implement logic for required courses having no valid schedules
+    for (Section section : courseSections) {
+      if (!sectionsOverlap(section, scheduleSections)
+          && currCredits + course.getCredits() <= invariants.getMaxCredits()) {
```

Comment:
This `if` statement is a bit hard to read. It might be helpful to assign this boolean expression to a variable named after the condition it's checking for, eg.

```
bool doesNotOverlapSchedule = !sectionsOverlap(....);
bool doesNotExceedCreditLimit = currCredits + ........;

if (doesNotOverlapSchedule && doesNotExceedCreditLimit) {
  ...
}
```

is easier to understand at a glance